### PR TITLE
feat: bundle and auto-load trusted setup

### PR DESCRIPTION
**Pull request by matthewkeil**
**Motivation**

Allows for easier initialization of library for Ethereum use cases.  Also bundles the trusted setup so that it does not need to be manually downloaded by consumers and stored separately.  Will simplify the `lodestar` use case.  Defaults to using the file passed by the user first.  If none is passed use the one found in the dist bundle (checks here for the prod case first).  If one is not found in the prod location fall back to the src location (for development purposes).  Still throws error if none is sourced (invalid bundle/build).

**Description of Changes**
- Bundle `trusted_setup.txt` with the c-kzg deps for node.js bindings
- Update `kzg.js` to functions to allow for passing a trusted setup or for loading the default setup from either the bundle or the repo src
- Export `getTrustedSetupFilepath` for testing purposes but do not announce via types file.
- Unit test `getTrustedSetupFilepath` to ensure proper loading.

### CODE REVIEW

File: bindings/node.js/lib/kzg.js [hunk 1]
```diff
@@ -6,6 +6,28 @@ const fs = require("fs");
...
+ * used for cases where the Ethereum official mainnet kzg setup is acceptable.
+ *
+ * @returns {string | undefined} - Filepath for trusted_setup.txt if found
+ */
+function getDefaultTrustedSetupFilepath() {
+  const locationsToSearch = [
+    // check the production bundle case first (this file in lib)
+    path.resolve(__dirname, "..", "deps", "c-kzg", "trusted_setup.txt"),
+    // check the development in-repo case second (this file in bindings/node.js/lib)
+    path.resolve(__dirname, "..", "..", "..", "src", "trusted_setup.txt"),
```
Comment (asn-d6) [inline bindings/node.js/lib/kzg.js:21]: Do we need this auto-detection logic? Why can't the user of the library always specify where the trusted setup is? Just worrying that the amount of `..` here makes this brittle.

What use cases are there, where someone loads a library and they don't know where the trusted setup file is, but the library knows?

File: bindings/node.js/README.md [hunk 2]
```diff
@@ -7,7 +7,7 @@ API. The core functionality was originally a stripped-down copy of
 since then. This package wraps that native `c-kzg` C code in C/C++ NAPI
 bindings for use in node.js applications.
 
-Spec: https://github.com/ethereum/consensus-specs/blob/dev/specs/eip4844/polynomial-commitments.md
+Spec: <https://github.com/ethereum/consensus-specs/blob/dev/specs/eip4844/polynomial-commitments.md>
```
Comment (jtraglia) [inline bindings/node.js/README.md:10]: Can we fix this link too?
```suggestion
Spec: <https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/polynomial-commitments.md>
```

File: bindings/node.js/lib/kzg.js [hunk 3]
```diff
@@ -31,19 +53,49 @@ function transformTrustedSetupJson(filePath) {
...
+ *
+ * @param {string} filePath - User provided filePath to check for trusted setup
+ *
+ * @returns {string} - Location of a trusted setup file. Validity is checked by
+ *                     the native bindings.loadTrustedSetup
+ *
+ * @throws {TypeError} - Invalid file type
+ * @throws {Error} - Invalid location or no default trusted setup found
+ *
+ * @remarks - This function is only exported for testing purposes.  It should
```
Comment (jtraglia) [inline bindings/node.js/lib/kzg.js:69]: Double space between sentences should be removed.
```suggestion
 * @remarks - This function is only exported for testing purposes. It should
```

File: bindings/node.js/lib/kzg.js [hunk 4]
```diff
@@ -31,19 +53,49 @@ function transformTrustedSetupJson(filePath) {
...
-bindings.loadTrustedSetup = function loadTrustedSetup(filePath) {
-  if (!(filePath && typeof filePath === "string")) {
-    throw new TypeError("must initialize kzg with the filePath to a txt/json trusted setup");
-  }
-  if (!fs.existsSync(filePath)) {
-    throw new Error(`no trusted setup found: ${filePath}`);
+/**
+ * Gets location for trusted setup file. Uses user provided location first. If
+ * one is not provided then defaults to the official Ethereum mainnet setup from
+ * the kzg ceremony.
```
Comment (jtraglia) [inline bindings/node.js/lib/kzg.js:59]: Nit: in the new code/comments, can we replace "kzg" with "KZG"? Except for "c-kzg" of course.
```suggestion
 * the KZG ceremony.
```

**Review by jtraglia — COMMENTED** [review]
This seems reasonable to me. I like that users can still specify a trusted setup.

File: bindings/node.js/README.md (hunk 2 above)
Comment (matthewkeil) [reply bindings/node.js/README.md:10]: Updated.

File: bindings/node.js/lib/kzg.js (hunk 4 above)
Comment (matthewkeil) [reply bindings/node.js/lib/kzg.js:59]: Updated it here and in all other locations "kzg" was in the docs and docstrings.

File: bindings/node.js/lib/kzg.js (hunk 3 above)
Comment (matthewkeil) [reply bindings/node.js/lib/kzg.js:69]: Updated.

**Comment by jtraglia** [comment]
@matthewkeil what does "enabling exceptions" do exactly and what are the implications?

File: bindings/node.js/lib/kzg.js (hunk 1 above)
Comment (matthewkeil) [reply bindings/node.js/lib/kzg.js]: > Do we need this auto-detection logic? Why can't the user of the library always specify where the trusted setup is? Just worrying that the amount of .. here makes this brittle.

Sadly yes.  This an unfortunate pattern that shows up often in JS repos.  Generally src and artifacts are found in different places and usually prod code runs with artifacts (using transpiled JS) and unit test use src file running with pure TypeScript code and  `ts-node` under the hood somewhere.

> What use cases are there, where someone loads a library and they don't know where the trusted setup file is, but the library knows?

If a consumer/user supplies a value and it is invalid the library will throw and will not load a default config.  This way the consumer will not end up inadvertently loading a setup they were not expecting to load.  The default versions are ONLY used if no location is passed to the `loadTrustedSetup` function.

File: bindings/node.js/binding.gyp [hunk 5]
```diff
@@ -28,15 +29,18 @@
           "sources": ["deps/blst/build/win64/*-x86_64.asm"],
           "defines": [
             "_CRT_SECURE_NO_WARNINGS",
+            "_HAS_EXCEPTIONS=1"
           ],
           "msbuild_settings": {
             "ClCompile": {
+              "ExceptionHandling": "Sync",
```
Comment (matthewkeil) [inline bindings/node.js/binding.gyp:36]: @jtraglia @asn-d6 I think this is the correct value here but would love a double check.  There is also an option for `SyncCThrow` but the docs recommend `Sync`.

https://learn.microsoft.com/en-us/visualstudio/msbuild/cl-task?view=vs-2022
https://learn.microsoft.com/en-us/cpp/build/reference/eh-exception-handling-model?view=msvc-170

File: bindings/node.js/lib/kzg.js (hunk 1 above)
Comment (asn-d6) [reply bindings/node.js/lib/kzg.js]: > @asn-d6 sadly yes. This an unfortunate pattern that shows up often in JS repos. Generally src and artifacts are found in different places and usually prod code runs with artifacts (using transpiled JS) and unit test use src file running with pure TypeScript code and ts-node under the hood somewhere.

Hm. It's still not clear to me why we can't have prod code and unit tests always pass the right file path to the `loadTrustedSetup` function. This way, we could remove this `getDefaultTrustedSetupFilepath()` function and simplify the code.

**Comment by jtraglia** [comment]
@matthewkeil thanks for the explanation. I think this is a bit over my head. Could you split the exception changes into a separate PR and ask a teammate (or just someone that would know) to review it?

**Comment by matthewkeil** [comment]
@jtraglia exceptions moved to #423

**Review by jtraglia — APPROVED** [review]
LGTM. Thanks for making those changes!

File: bindings/node.js/lib/kzg.js (hunk 1 above)
Comment (matthewkeil) [reply bindings/node.js/lib/kzg.js]: I will see what I can do to simplify things.  This is how I've learned to do an how I've always done it.  Will have to double check that I catch all the edge cases to make sure it works everywhere.  Stay-tuned @asn-d6.  Will try to get this pushed tomorrow morning (is almost midnight here).

Comment (matthewkeil) [reply bindings/node.js/lib/kzg.js]: Will do!!  Sorry @asn-d6 will get this up asap

Comment (matthewkeil) [reply bindings/node.js/lib/kzg.js]: Updated in https://github.com/ethereum/c-kzg-4844/pull/424

**Pull request merged**